### PR TITLE
feat(sql): ceil() and floor() functions

### PR DESCRIPTION
**Pull request by mrecht**
Implement ceil() and floor() math functions for Double and Float types.

### CODE REVIEW

**Comment by puzpuzpuz** [comment]
There are existing numeric functions like `round*()` (see this [doc section](https://questdb.io/docs/reference/function/numeric)) supported by QuestDB.

For positive numbers `round_down()` seems to be doing what `Math#floor` does and `round_up()` does the same thing as `Math#floor`.

The behavior for negative numbers is different, but I wonder if that difference is worth it to add new numeric functions with similar functionality? Could you describe the use case where you need these `ceil()` and `floor()` functions?

**Comment by mrecht** [comment]
> There are existing numeric functions like `round*()` (see this [doc section](https://questdb.io/docs/reference/function/numeric)) supported by QuestDB.
> 
> For positive numbers `round_down()` seems to be doing what `Math#floor` does and `round_up()` does the same thing as `Math#floor`.
> 
> The behavior for negative numbers is different, but I wonder if that difference is worth it to add new numeric functions with similar functionality? Could you describe the use case where you need these `ceil()` and `floor()` functions?

`ceil()` and `floor()` are part of std. SQL and as you described behave different with negative numbers. In my opinion adding the two functions make transitioning from other SQL based systems easier.

**Comment by puzpuzpuz** [comment]
> `ceil()` and `floor()` are part of std. SQL and as you described behave different with negative numbers. In my opinion adding the two functions make transitioning from other SQL based systems easier.

Thanks, it makes sense.

File: core/src/test/java/io/questdb/griffin/engine/functions/math/CeilDoubleFunctionFactoryTest.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+    @Test
+    public void testNegative() throws SqlException {
+        call(-13.1).andAssert(-13.0, 0.0000000001);
+    }
+
+    @Override
+    protected FunctionFactory getFunctionFactory() { return new CeilDoubleFunctionFactory();
+    }
+}
```
Comment (puzpuzpuz) [inline core/src/test/java/io/questdb/griffin/engine/functions/math/CeilDoubleFunctionFactoryTest.java:47]: nit: a trailing new line is missing here and in some other class files.

File: core/src/test/java/io/questdb/griffin/engine/functions/math/CeilFloatFunctionFactoryTest.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import io.questdb.griffin.FunctionFactory;
+import io.questdb.griffin.SqlException;
+import io.questdb.griffin.engine.AbstractFunctionFactoryTest;
+import org.junit.Test;
+
+public class CeilFloatFunctionFactoryTest extends AbstractFunctionFactoryTest {
+
+    @Test
+    public void testPositive() throws SqlException {
+        // call(13.1f).andAssert(14.0, 0.0000000001);
```
Comment (puzpuzpuz) [inline core/src/test/java/io/questdb/griffin/engine/functions/math/CeilFloatFunctionFactoryTest.java:36]: This looks like a leftover.

File: core/src/test/java/io/questdb/griffin/engine/functions/math/CeilFloatFunctionFactoryTest.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+ ******************************************************************************/
+
+package io.questdb.griffin.engine.functions.math;
+
+import io.questdb.griffin.FunctionFactory;
+import io.questdb.griffin.SqlException;
+import io.questdb.griffin.engine.AbstractFunctionFactoryTest;
+import org.junit.Test;
+
+public class CeilFloatFunctionFactoryTest extends AbstractFunctionFactoryTest {
```
Comment (puzpuzpuz) [inline core/src/test/java/io/questdb/griffin/engine/functions/math/CeilFloatFunctionFactoryTest.java:32]: Would be nice to include tests for `Double.NaN`/`Float.NaN` into this and other test suites.

File: core/src/test/java/io/questdb/griffin/engine/functions/math/CeilDoubleFunctionFactoryTest.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        call(13.1).andAssert(14.0, 0.0000000001);
+    }
+
+    @Test
+    public void testNegative() throws SqlException {
+        call(-13.1).andAssert(-13.0, 0.0000000001);
+    }
+
+    @Override
+    protected FunctionFactory getFunctionFactory() { return new CeilDoubleFunctionFactory();
```
Comment (puzpuzpuz) [inline core/src/test/java/io/questdb/griffin/engine/functions/math/CeilDoubleFunctionFactoryTest.java:45]: `return new CeilDoubleFunctionFactory();` should start on a new line. The same suggestion applies to other new tests.

File: core/src/test/java/io/questdb/griffin/engine/functions/math/FloorFloatFunctionFactoryTest.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+import io.questdb.griffin.FunctionFactory;
+import io.questdb.griffin.SqlException;
+import io.questdb.griffin.engine.AbstractFunctionFactoryTest;
+import org.junit.Test;
+
+public class FloorFloatFunctionFactoryTest extends AbstractFunctionFactoryTest {
+
+    @Test
+    public void testPositive() throws SqlException {
+        // call(13.1f).andAssert(13.0, 0.0000000001);
```
Comment (puzpuzpuz) [inline core/src/test/java/io/questdb/griffin/engine/functions/math/FloorFloatFunctionFactoryTest.java:36]: Another leftover, probably.

File: core/src/main/java/io/questdb/griffin/engine/functions/math/CeilDoubleFunctionFactory.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        return "ceil(D)";
+    }
+
+    @Override
+    public Function newInstance(int position, ObjList<Function> args, IntList argPositions, CairoConfiguration configuration, SqlExecutionContext sqlExecutionContext) {
+        return new CeilFunction(args.getQuick(0));
+    }
+
+    private static class CeilFunction extends DoubleFunction implements UnaryFunction {
+        final Function function;
```
[the remark this answers is not in the document]
Comment (mrecht) [reply core/src/main/java/io/questdb/griffin/engine/functions/math/CeilDoubleFunctionFactory.java]: done

File: core/src/test/java/io/questdb/griffin/engine/functions/math/CeilFloatFunctionFactoryTest.java (hunk 2 above)
Comment (mrecht) [reply core/src/test/java/io/questdb/griffin/engine/functions/math/CeilFloatFunctionFactoryTest.java]: done

File: core/src/test/java/io/questdb/griffin/engine/functions/math/FloorFloatFunctionFactoryTest.java (hunk 5 above)
Comment (mrecht) [reply core/src/test/java/io/questdb/griffin/engine/functions/math/FloorFloatFunctionFactoryTest.java]: removed

File: core/src/test/java/io/questdb/griffin/engine/functions/math/CeilDoubleFunctionFactoryTest.java (hunk 1 above)
Comment (mrecht) [reply core/src/test/java/io/questdb/griffin/engine/functions/math/CeilDoubleFunctionFactoryTest.java]: done

File: core/src/test/java/io/questdb/griffin/engine/functions/math/CeilFloatFunctionFactoryTest.java (hunk 3 above)
Comment (mrecht) [reply core/src/test/java/io/questdb/griffin/engine/functions/math/CeilFloatFunctionFactoryTest.java:32]: added tests for NaN

**Review by bluestreak01 — APPROVED** [review]

**Comment by bluestreak01** [comment]
perfect, thank you @mrecht and @puzpuzpuz 👍

**Comment by puzpuzpuz** [comment]
@mrecht thanks for the contribution!

**Pull request merged**